### PR TITLE
add parallel build script to ci …

### PR DESCRIPTION
**Pull request by wopeizl**
… test=develop
1. 将多卡隔离为多个单卡/双卡环境，来并行跑多个单测
2. 单测分类，将单测分为可单卡运行/双卡运行/独占运行三种类型。

4.1日CI单测时间（617个单测）： 4卡机器 56 分钟，8卡机器 53 分钟
4.19日CI单测时间（617个单测）：4卡机器 33 分钟，8卡机器 27 分钟 

部分独占运行case列表如下，
 1/24 Test #146: test_inference_label_semantic_roles ..........   Passed   11.74 sec
 2/24 Test #142: test_api_impl ................................   Passed   15.35 sec
 3/24 Test #144: test_inference_image_classification_vgg ......   Passed   17.29 sec
 4/24 Test #145: test_inference_image_classification_resnet ...   Passed   17.39 sec
 5/24 Test #147: test_inference_recognize_digits_mlp ..........   Passed    8.66 sec
 6/24 Test #151: test_inference_nlp ...........................   Passed    2.11 sec
 7/24 Test #149: test_inference_recommender_system ............   Passed    8.33 sec
 8/24 Test #174: test_train_recognize_digits_mlp ..............   Passed    2.88 sec
 9/24 Test #150: test_inference_word2vec ......................   Passed    8.58 sec
10/24 Test #148: test_inference_recognize_digits_conv .........   Passed   14.64 sec
11/24 Test #307: test_hsigmoid_remote_table_op ................   Passed    3.21 sec
12/24 Test #252: test_parallel_executor_test_while_train ......   Passed    7.83 sec
13/24 Test #365: test_nce_remote_table_op .....................   Passed    3.63 sec
14/24 Test #175: test_train_recognize_digits_conv .............   Passed   16.14 sec
15/24 Test #384: test_listen_and_serv_op ......................   Passed    4.85 sec
16/24 Test #438: test_alloc_continuous_space_op ...............   Passed    5.71 sec
17/24 Test #337: test_parallel_executor_mnist .................   Passed   21.12 sec
18/24 Test #445: test_conv_shift_op ...........................   Passed    5.49 sec
19/24 Test #541: test_adam_op_multi_thread ....................   Passed    5.99 sec
20/24 Test #205: test_weight_decay ............................   Passed   62.28 sec
21/24 Test #427: test_recordio_reader .........................   Passed   56.72 sec
22/24 Test #452: test_parallel_executor_seresnext .............   Passed  428.13 sec
23/24 Test #544: test_nearest_interp_op .......................   Passed   58.07 sec
24/24 Test #552: test_parallel_executor_crf ...................   Passed   81.98 sec

部分双卡case，
 Test #255: test_dist_base ...................   Passed    2.23 sec
 Test #192: test_distribute_fpn_proposals_op ...   Passed    6.51 sec
 Test #267: test_dist_save_load ..............   Passed   24.87 sec
 Test #410: test_dist_allreduce_op .............   Passed   25.29 sec
 Test #232: test_dist_mnist_pg ...............   Passed   25.48 sec
 Test #278: test_dist_mnist_batch_merge ......   Passed   19.90 sec
 Test #614: test_distillation_strategy .......   Passed   62.71 sec
 Test #275: test_dist_simnet_bow ...............   Passed   47.75 sec
 Test #319: test_dist_ctr ....................   Passed   26.57 sec
 Test #492: test_dist_word2vec ...............   Passed   66.63 sec
 Test #531: test_dist_text_classification ....   Passed   31.28 sec
 Test #358: test_dist_mnist ..................   Passed  125.71 sec
 Test #548: test_dist_train ..................   Passed    2.59 sec
 Test #551: test_dist_se_resnext_nccl ........   Passed   60.43 sec
 Test #550: test_dist_se_resnext ...............   Passed  233.77 sec

### CODE REVIEW

**Review by junjun315 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: paddle/scripts/paddle_build.sh [hunk 1]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
...
+                # echo "env CUDA_VISIBLE_DEVICES=$cuda_list ctest -I $i,,$NUM_PROC -R \"($testcases)\" --output-on-failure &"
+                env CUDA_VISIBLE_DEVICES=$cuda_list ctest -I $i,,$NUM_PROC -R "($testcases)" --output-on-failure &
+            fi
+        fi
+    done
+
+    wait; # wait for all subshells to finish
+}
+
+function aggresive_test() {
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:680]: aggresive没有这个单词

File: paddle/scripts/paddle_build.sh [hunk 2]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
...
+    Running unit tests ...
+    ========================================
+EOF
+
+set +x
+        EXIT_CODE=0;
+        test_cases=$(ctest -N -V)
+        exclusive_tests=''
+        single_card_tests=''
+        multiple_card_tests=''
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:695]: 请解释下每个变量的含义
```
exclusive_tests='' 这个没理解是什么测试？独占的？
single_card_tests='' 单卡测试
multiple_card_tests='' 多卡测试
```

File: paddle/scripts/paddle_build.sh [hunk 3]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
...
+        is_multicard=''
+        while read -r line; do
+            if [[ "$line" == "" ]]; then
+                continue
+            fi
+                read matchstr <<< $(echo "$line"|grep -oEi 'Test[ \t]+#')
+                if [[ "$matchstr" == "" ]]; then
+                    # Any test case with LABELS property would be parse here
+                    # RUN_TYPE=EXCLUSIVE mean the case would run exclusively
+                    # RUN_TYPE=DIST mean the case would take two graph cards during runtime
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:706]: 706行的解释是说，分布式的单测都必须要2张卡，4张卡都不行？

File: paddle/scripts/paddle_build.sh [hunk 4]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
...
+                    read is_exclusive <<< $(echo "$line"|grep -oEi "RUN_TYPE=EXCLUSIVE")
+                    read is_multicard <<< $(echo "$line"|grep -oEi "RUN_TYPE=DIST")
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:708]: `RUN_TYPE=EXCLUSIVE` 和`SERIAL`重复了。

File: paddle/scripts/paddle_build.sh [hunk 5]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
     wait
 }
 
+function parallel_test() {
+    mkdir -p ${PADDLE_ROOT}/build
+    cd ${PADDLE_ROOT}/build
+    if [ ${WITH_TESTING:-ON} == "ON" ] ; then
+    cat <<EOF
+    ========================================
+    Running unit tests ...
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:570]: Running unit tests ... 没体现并行

File: paddle/scripts/paddle_build.sh [hunk 6]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
     wait
 }
 
+function parallel_test() {
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:564]: 这个`parallel_test`没有看到用的地方。可以删掉？

File: paddle/scripts/paddle_build.sh [hunk 7]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
...
+            if [[ $cardnumber == $CUDA_DEVICE_COUNT ]]; then
+                ctest -I $i,,$NUM_PROC -R "($testcases)" -V &
+            else
+                env CUDA_VISIBLE_DEVICES=$cuda_list ctest -I $i,,$NUM_PROC -R "($testcases)" -V &
+            fi
+        else
+            if [[ $cardnumber == $CUDA_DEVICE_COUNT ]]; then
+                ctest -I $i,,$NUM_PROC -R "($testcases)" --output-on-failure &
+            else
+                # echo "env CUDA_VISIBLE_DEVICES=$cuda_list ctest -I $i,,$NUM_PROC -R \"($testcases)\" --output-on-failure &"
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:671]: 671行可以去掉

File: paddle/scripts/paddle_build.sh [hunk 8]
```diff
@@ -561,6 +561,195 @@ function bind_test() {
...
+        # ctest -I https://cmake.org/cmake/help/v3.0/manual/ctest.1.html?highlight=ctest
+        cuda_list=()
+        for (( j = 0; j < cardnumber; j++ )); do
+            if [ $j -eq 0 ]; then
+                    cuda_list=("$[i*cardnumber]")
+                else
+                    cuda_list="$cuda_list,$[i*cardnumber+j]"
+            fi
+        done
+        # echo $cuda_list
```
Comment (luotao1) [inline paddle/scripts/paddle_build.sh:660]: 660行可以去掉

File: paddle/scripts/paddle_build.sh (hunk 7 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:671]: Thanks, trigger a new pr: https://github.com/PaddlePaddle/Paddle/pull/17020

File: paddle/scripts/paddle_build.sh (hunk 8 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:660]: Thanks, trigger a new pr: https://github.com/PaddlePaddle/Paddle/pull/17020

File: paddle/scripts/paddle_build.sh (hunk 6 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:564]: Thanks, trigger a new pr: https://github.com/PaddlePaddle/Paddle/pull/17020

File: paddle/scripts/paddle_build.sh (hunk 5 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:570]: Thanks, trigger a new pr: https://github.com/PaddlePaddle/Paddle/pull/17020

File: paddle/scripts/paddle_build.sh (hunk 3 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:706]: 一般情况下两卡应该是够

File: paddle/scripts/paddle_build.sh (hunk 1 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:680]: Thanks, trigger a new pr: https://github.com/PaddlePaddle/Paddle/pull/17020

File: paddle/scripts/paddle_build.sh (hunk 2 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:695]: Thanks, trigger a new pr: https://github.com/PaddlePaddle/Paddle/pull/17020

File: paddle/scripts/paddle_build.sh (hunk 4 above)
Comment (wopeizl) [reply paddle/scripts/paddle_build.sh:708]: 现在含义不太一样了，SERIAL在现在处理方式下应该不起作用了

**Pull request merged**